heracles/dices/shrinkage.py

```diff
@@ -103,76 +98,55 @@ def shrinkage_factor(cls1, target):
+        cl1 = cls[key1]
+        cl2 = cls[key2]
         # get attributes of result
-        ell1 = get_result_array(cl1, "ell")
-        ell2 = get_result_array(cl2, "ell")
-        ell = ell1 + ell2
-        r = np.zeros(
-            (len(idx1), len(idx2), len(idx3), len(idx4), len(ell1[0]), len(ell2[0]))
-        )
-        # get covariance
-        for k, idx in zip(
-            itertools.product(_a1, _b1, _a2, _b2),
-            itertools.product(idx1, idx2, idx3, idx4),
-        ):
-            __a1, __b1, __a2, __b2 = k
-            _key = (__a1, __b1, __a2, __b2, i1, j1, i2, j2)
-            _cov = _gaussian_covariance(_Cls, _key)
-            ix1, ix2, ix3, ix4 = idx
-            r[ix1, ix2, ix3, ix4, :, :] = np.diag(_cov)
-        # Remove the extra dimensions
-        r = np.squeeze(r)
-        # Make Result
-        result = Result(r, spin=(sa1, sb1, sa2, sb2), ell=ell)
-        cov[covkey] = result
-    return cov
-
-
-def _gaussian_covariance(cls, key):
-    """
-    Returns a particular entry of the gaussian covariance matrix.
-    input:
-        cls: Cls
-        key: key of the entry
-    returns:
-        cov: covariance matrix
-    """
-    a1, b1, a2, b2, i1, j1, i2, j2 = key
-    cl1 = get_cl((a1, a2, i1, i2), cls)
-    cl2 = get_cl((b1, b2, j1, j2), cls)
-    cl3 = get_cl((a1, b2, i1, j2), cls)
-    cl4 = get_cl((b1, a2, j1, i2), cls)
-    cov = cl1 * cl2 + cl3 * cl4
+        (ell1,) = get_result_array(cl1, "ell")
+        (ell2,) = get_result_array(cl2, "ell")
+        # get dof by expanding spin0 dims
+        # Eg. POS POS goes from (l) to (1, 1, l)
+        cl1 = expand_spin0_dims(cls[key1])
+        cl2 = expand_spin0_dims(cls[key2])
+        print("ell1:", ell1, "ell2:", ell2)
+        # keys for cov
+        _cl1 = get_cl((a1, a2, i1, i2), cls)
+        _cl2 = get_cl((b1, b2, j1, j2), cls)
+        _cl3 = get_cl((a1, b2, i1, j2), cls)
+        _cl4 = get_cl((b1, a2, j1, i2), cls)
+        # get spins
+        _cl1 = expand_spin0_dims(_cl1)
+        _cl2 = expand_spin0_dims(_cl2)
+        _cl3 = expand_spin0_dims(_cl3)
+        _cl4 = expand_spin0_dims(_cl4)
+        # shape of the result
+        dof_a1, dof_b1, _ = cl1.shape
+        dof_a2, dof_b2, _ = cl2.shape
+        _ell = min(len(ell1), len(ell2))
+        print("dofs: ", dof_a1, dof_b1, dof_a2, dof_b2, _ell)
+        r = np.zeros((dof_a1, dof_b1, dof_a2, dof_b2, _ell))
+        print(r.shape)
+        for _1, _2, _3, _4 in np.ndindex(r.shape[:-1]):
+            _r = _cl1[_1, _3] * _cl2[_2, _4] + _cl3[_1, _4] * _cl4[_2, _3]
+            print("_r shape: ", _r.shape, "r shape:", r[_1, _2, _3, _4, :].shape)
+            r[_1, _2, _3, _4, :] = _r
+        # make diagonal into matrix
+        eye = np.eye(r.shape[-1])
+        r = r[..., :, None] * eye
+        # Assign to cov
+        _ax = np.arange(len(r.shape))
+        ax1, ax2 = int(_ax[-2]), int(_ax[-1])
+        r = Result(r, spin=(*cl1.spin, *cl2.spin), ell=(ell1, ell2), axis=(ax1, ax2))
```

Comment:
Result does the right thing internally
```suggestion
        r = Result(r, spin=(*cl1.spin, *cl2.spin), ell=(ell1, ell2), axis=(-2, -1))
```